### PR TITLE
Drill 3952

### PR DESCRIPTION
**Pull request by adeneche**
Improve Window Functions performance when not all batches are required to process the current batch

This patch also improves how the window functions operator handles limits by avoiding processing anymore batches after the kill signal has been received

@amansinha100 can you please review ?

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java [hunk 1]
```diff
@@ -287,6 +318,36 @@ private int countPeers(final int start) {
...
 
   /**
+   * Counts how many rows are peer with the first row of the current frame. This is called when we don't require all
+   * batches of current partition to be processed at once.<br>
+   * Assumes the end of the frame has indeed been found, because of this it doesn't use partition.remaining to check
+   * the end of current partition but rather isSamePartition()
+   * @param start first row of current frame
+   * @return number of peer rows
+   */
+  private int countPeersPartialPartition(final int start) {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: This should be a long since the number of peers could easily be greater than Integer.MAX_VALUE.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java [hunk 2]
```diff
@@ -287,6 +318,36 @@ private int countPeers(final int start) {
...
+
+    // count all rows that are in the same frame of starting row
+    // keep increasing length until we find first non peer row we reach the very
+    // last batch
+    for (WindowDataBatch batch : batches) {
+      final int recordCount = batch.getRecordCount();
+
+      // for every remaining row in the partition, count it if it's a peer row
+      for (int row = (batch == first) ? start : 0; row < recordCount; row++, length++) {
+        if (!isSamePartition(start, first, row, batch) || !isPeer(start, first, row, batch)) {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: I think isPeer() could also check whether two rows are in the same partition...because the notion of peer is only valid within a partition.  This will simplify the generated code a bit.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java [hunk 3]
```diff
@@ -287,6 +318,36 @@ private int countPeers(final int start) {
...
+    int length = 0;
+
+    // count all rows that are in the same frame of starting row
+    // keep increasing length until we find first non peer row we reach the very
+    // last batch
+    for (WindowDataBatch batch : batches) {
+      final int recordCount = batch.getRecordCount();
+
+      // for every remaining row in the partition, count it if it's a peer row
+      for (int row = (batch == first) ? start : 0; row < recordCount; row++, length++) {
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: The performance could be pretty bad due to the sequential scan through the remaining rows.  The data is sorted right ?

**Comment by amansinha100** [comment]
Overall,  there are 2 categories of the window functions: one that require all batches of the partition before they can start processing and second that can start work with partial partition as long as we have all peer rows of the current row.   It seems to me that some refactoring is needed to have derived classes dedicated to the 2 categories, otherwise  the logic in the common code is getting complicated.  What do you think ?  This need not necessarily be done as part of this patch...but we should discuss.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java (hunk 2 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: why not. We still need to have a separate `isSamePartition` though

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java (hunk 1 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: will fix

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java (hunk 3 above)
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/DefaultFrameTemplate.java]: This may be a relic from early implementations. I should be able to remove it now.

I may also be able to get rid of `updatePartitionSize()` but this may require too much work for this patch and may do it in a separate patch

**Comment by adeneche** [comment]
Window operator has become increasingly complex. I already opened [DRILL-3662](https://issues.apache.org/jira/browse/DRILL-3662) to refactor the code.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/WindowFramer.java [hunk 4]
```diff
@@ -38,19 +40,33 @@ void setup(final List<WindowDataBatch> batches, final VectorContainer container,
...
+                                          @Named("b2Index") int b2Index, @Named("b2") VectorAccessible b2);
+
+  /**
+   * compares two rows from different batches (can be the same), if they have the same value for the order by
+   * expression
+   * @param b1Index index of first row
+   * @param b1 batch for first row
+   * @param b2Index index of second row
+   * @param b2 batch for second row
+   * @return true if the rows are in the same partition
```
Comment (amansinha100) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/window/WindowFramer.java]: change to 'rows in the same partition and are peers'

**Comment by amansinha100** [comment]
+1.   Per previous comments, there is room for improvement in terms of code organization and reducing the number of passes which I understand will be covered by DRILL-3662.

**Comment by adeneche** [comment]
thanks @amansinha100 , I updated DRILL-3662 description with the new refactoring tasks
